File: pyoasis/examples/14-test-inter3d/python/sender-cube.py
#!/usr/bin/env python3

import os
import math
import numpy as np
import pyoasis
from pyoasis import OASIS
import netCDF4
from mpi4py import MPI

comm = MPI.COMM_WORLD

has_graphics = None
has_graphics = comm.bcast(has_graphics, root=comm.size - 1)

sgrid = 'nogt'
ll_plot = None
if comm.rank == 0:
    do_plot = input('Plot output [yes/no]\n')
    if (do_plot.lower() != 'yes' and do_plot.lower() != 'no'):
        print('{} is not a valid yes/no answer'.format(do_plot), flush=True)
        comm.Abort()
    else:
        ll_plot = do_plot.lower() == 'yes'

ll_plot = comm.bcast(ll_plot, root=0)

component_name = "sender-cube"
comp = pyoasis.Component(component_name, True, comm)

if comm.rank == 0:
    print(comp, flush=True)

comm_rank = comp.localcomm.rank
comm_size = comp.localcomm.size

sgrid = comp.localcomm.bcast(sgrid, root=0)

gf = netCDF4.Dataset('grids.nc', 'r')
lons = gf.variables[sgrid + '.lon'][:, :].data.T
lats = gf.variables[sgrid + '.lat'][:, :].data.T

nx_global = gf.dimensions['x_'+sgrid].size
ny_global = gf.dimensions['y_'+sgrid].size
nz_global = 3
n_points_2d = nx_global * ny_global
n_points_3d = n_points_2d * nz_global
gf.close()

if comm_rank == 0:
    print(comp)
    print("n_points per level on source side is {}".format(n_points_2d))
    print("n_levs on source side is {}".format(nz_global))

xdec = math.ceil(math.sqrt(comm_size))
while xdec <= comm_size:
    if comm_size % xdec == 0:
        break
ydec = int(comm_size/xdec)
nx = int(nx_global / xdec)
ny = int(ny_global / ydec)
idx = nx * (comm_rank % xdec)
idy = ny * math.floor(float(comm_rank)/float(xdec))
if idx + nx > nx_global:
    nx = nx_global - idx
if idy + ny > ny_global:
    ny = ny_global - idy
offset = nx_global * idy + idx

partitionb = pyoasis.BoxPartition(offset, nx, ny, nx_global, global_size=n_points_2d)
partitionc = pyoasis.CubePartition(offset, nx, ny, nx_global, ny_global, nz_global,
                                   global_size=n_points_3d)

fsendlv1_var = pyoasis.Var("FSENDLV1", partitionb, OASIS.OUT)
fsendlv2_var = pyoasis.Var("FSENDLV2", partitionb, OASIS.OUT)
fsendlv3_var = pyoasis.Var("FSENDLV3", partitionb, OASIS.OUT)
fsendf3d_var = pyoasis.Var("FSENDF3D", partitionc, OASIS.OUT)
comp.enddef()

local_size = nx * ny
date = int(0)
field = pyoasis.asarray(np.zeros((nx,ny,nz_global), dtype=np.float64))

dp_conv = math.pi / 180.
field[:,:,0] = 2.0 + np.sin(4.0 * lats[idx:idx+nx,idy:idy+ny] * dp_conv) ** 4.0 * \
               np.cos(8.0 * lons[idx:idx+nx,idy:idy+ny] * dp_conv)

field[:,:,1] = 2.0 + np.sin(2.0 * lats[idx:idx+nx,idy:idy+ny] * dp_conv) ** 4.0 * \
               np.cos(4.0 * lons[idx:idx+nx,idy:idy+ny] * dp_conv)

field[:,:,2] = 2.0 - np.cos(math.pi *
                            (np.arccos(np.cos(lons[idx:idx+nx,idy:idy+ny] * dp_conv) *
                                       np.cos(lats[idx:idx+nx,idy:idy+ny] * dp_conv)) /
                             (1.2 * math.pi)))

if comm_rank == 0:
    print("Sent data: at time {}".format(date))

fsendlv1_var.put(date, field[:,:,0])
fsendlv2_var.put(date, field[:,:,1])
fsendlv3_var.put(date, field[:,:,2])
fsendf3d_var.put(date, field)

del comp

File: pyoasis/examples/14-test-inter3d/C/sender-cube.c
#include <stdio.h>
#include <math.h>
#include "mpi.h"
#include "netcdf.h"
#include "oasis_c.h"

int main(int argc, char *argv[])
{
  char *comp_name = "sender-cube";
  int comp_id;
  OASIS_CHECK_ERR(oasis_c_init_comp(&comp_id, comp_name, OASIS_COUPLED));

  MPI_Comm localcomm;
  OASIS_CHECK_ERR(oasis_c_get_localcomm(&localcomm));
  int comm_size;
  OASIS_CHECK_MPI_ERR(MPI_Comm_size(localcomm, &comm_size));
  int comm_rank;
  OASIS_CHECK_MPI_ERR(MPI_Comm_rank(localcomm, &comm_rank));
  if (comm_rank == 0) {
    fprintf(stdout,"%s: Component ID: %d\n", comp_name, comp_id);
  }

  int nx_global = 362;
  int ny_global = 294;
  int nz_global = 3;
  int n_points = nx_global*ny_global;

  int xdec;
  for (xdec = ceil(sqrt((float)comm_size)); xdec <= comm_size; xdec++) {
    if (comm_size % xdec == 0) break;
  }
  int ydec = comm_size / xdec;

  int ncid, varid;
  if (nc_open("grids.nc", 0, &ncid) != NC_NOERR)
    OASIS_CHECK_ERR(oasis_c_abort(comp_id, comp_name, "Sender: Error in opening grids.nc",
				    __FILE__, __LINE__));

  float lon[ny_global][nx_global];
  float lat[ny_global][nx_global];
  if (nc_inq_varid(ncid, "nogt.lon", &varid) != NC_NOERR)
    OASIS_CHECK_ERR(oasis_c_abort(comp_id, comp_name, "Sender: Error in getting nogt.lon id",__FILE__, __LINE__));
  if (nc_get_var(ncid, varid, &lon[0][0]) != NC_NOERR)
    OASIS_CHECK_ERR(oasis_c_abort(comp_id, comp_name, "Sender: Error in getting nogt.lon values",__FILE__, __LINE__));
  if (nc_inq_varid(ncid, "nogt.lat", &varid) != NC_NOERR)
    OASIS_CHECK_ERR(oasis_c_abort(comp_id, comp_name, "Sender: Error in getting nogt.lat id",__FILE__, __LINE__));
  if (nc_get_var(ncid, varid, &lat[0][0]) != NC_NOERR)
    OASIS_CHECK_ERR(oasis_c_abort(comp_id, comp_name, "Sender: Error in getting nogt.lat values",__FILE__, __LINE__));
  if (nc_close(ncid))
    OASIS_CHECK_ERR(oasis_c_abort(comp_id, comp_name, "Sender: Error in closing grids.nc",__FILE__, __LINE__));

  int nx = nx_global / xdec;
  int ny = ny_global / ydec;

  int idx = nx * (comm_rank % xdec);
  int idy = ny * floor((float)comm_rank / (float)xdec);

  if ( idx + nx > nx_global ) nx = nx_global - idx;
  if ( idy + ny > ny_global ) ny = ny_global - idx;

  int offset = nx_global * idy + idx;

  int partb_params[OASIS_Box_Params];
  partb_params[OASIS_Strategy] = OASIS_Box;
  partb_params[OASIS_Offset]   = offset;
  partb_params[OASIS_SizeX]    = nx;
  partb_params[OASIS_SizeY]    = ny;
  partb_params[OASIS_LdX]      = nx_global;
  int partb_id;
  OASIS_CHECK_ERR(oasis_c_def_partition(&partb_id, OASIS_Box_Params,
					partb_params, nx_global * ny_global,
					OASIS_No_Name));

  int partc_params[OASIS_Cube_Params];
  partc_params[OASIS_Strategy] = OASIS_Cube;
  partc_params[OASIS_Offset]   = offset;
  partc_params[OASIS_SizeX]    = nx;
  partc_params[OASIS_SizeY]    = ny;
  partc_params[OASIS_LdX]      = nx_global;
  partc_params[OASIS_LdY]      = ny_global;
  partc_params[OASIS_LdZ]      = nz_global;
  int partc_id;
  OASIS_CHECK_ERR(oasis_c_def_partition(&partc_id, OASIS_Cube_Params,
					partc_params, nx_global * ny_global * nz_global,
					OASIS_No_Name));


  char *fl1_name  = "FSENDLV1";
  char *fl2_name  = "FSENDLV2";
  char *fl3_name  = "FSENDLV3";
  char *f3d_name  = "FSENDF3D";
  int bundle_size = 1;
  int fl1_id, fl2_id, fl3_id, f3d_id;
  OASIS_CHECK_ERR(oasis_c_def_var(&fl1_id, fl1_name, partb_id, bundle_size, OASIS_OUT, OASIS_DOUBLE));
  OASIS_CHECK_ERR(oasis_c_def_var(&fl2_id, fl2_name, partb_id, bundle_size, OASIS_OUT, OASIS_DOUBLE));
  OASIS_CHECK_ERR(oasis_c_def_var(&fl3_id, fl3_name, partb_id, bundle_size, OASIS_OUT, OASIS_DOUBLE));
  OASIS_CHECK_ERR(oasis_c_def_var(&f3d_id, f3d_name, partc_id, bundle_size, OASIS_OUT, OASIS_DOUBLE));
  OASIS_CHECK_ERR(oasis_c_enddef());

  int local_size = (int) n_points / comm_size ;
  if ( comm_rank == comm_size - 1) local_size = n_points - offset;

  double dp_conv = atan(1.)/45.0;
  double field[nz_global][ny][nx];
  int i, j, ll_i, ll_j;
  for (i = 0; i<nx; i++) {
    for (j = 0; j<ny; j++) {
      ll_i = idx + i;
      ll_j = idy + j;
      field[0][j][i] = 2.0 + pow(sin(4.0*lat[ll_j][ll_i]*dp_conv),4.0) *
	cos(8.0 * lon[ll_j][ll_i]*dp_conv);
      field[1][j][i] = 2.0 + pow(sin(2.0*lat[ll_j][ll_i]*dp_conv),4.0) *
	cos(4.0 * lon[ll_j][ll_i]*dp_conv);
      field[2][j][i] = 2.0 - cos(atan(1.)*4.*
			     (acos(cos(lon[ll_j][ll_i]*dp_conv)*cos(lat[ll_j][ll_i]*dp_conv))/
			      (1.2*atan(1.)*4)));
    }
  }
  int date = 0;

  int kinfo;
  OASIS_CHECK_ERR(oasis_c_put(fl1_id, date, nx * ny, 1, bundle_size, OASIS_DOUBLE, OASIS_COL_MAJOR, &field[0][0][0], OASIS_No_Restart, &kinfo));
  fprintf(stdout, "Sender rank(%d): oasis_c_put at time %d returned kinfo = %d\n", comm_rank, date, kinfo);
  fflush(stdout);
  OASIS_CHECK_ERR(oasis_c_put(fl2_id, date, nx * ny, 1, bundle_size, OASIS_DOUBLE, OASIS_COL_MAJOR, &field[1][0][0], OASIS_No_Restart, &kinfo));
  fprintf(stdout, "Sender rank(%d): oasis_c_put at time %d returned kinfo = %d\n", comm_rank, date, kinfo);
  fflush(stdout);
  OASIS_CHECK_ERR(oasis_c_put(fl3_id, date, nx * ny, 1, bundle_size, OASIS_DOUBLE, OASIS_COL_MAJOR, &field[2][0][0], OASIS_No_Restart, &kinfo));
  fprintf(stdout, "Sender rank(%d): oasis_c_put at time %d returned kinfo = %d\n", comm_rank, date, kinfo);
  fflush(stdout);
  OASIS_CHECK_ERR(oasis_c_put(f3d_id, date, nx * ny * nz_global, 1, bundle_size, OASIS_DOUBLE, OASIS_COL_MAJOR, field, OASIS_No_Restart, &kinfo));
  fprintf(stdout, "Sender rank(%d): oasis_c_put at time %d returned kinfo = %d\n", comm_rank, date, kinfo);
  fflush(stdout);

  OASIS_CHECK_ERR(oasis_c_terminate());

}
